src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -53,33 +53,48 @@ public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilatio
                 return this;
             }
 
-            // PERF: if the input compilation is the same as the last compilation we saw, and we have a final compilation
-            //       we know nothing can have changed and can just short circuit, returning the already processed final compilation 
-            if (compilation == _state.Compilation && _state.FinalCompilation is object)
-            {
-                outputCompilation = _state.FinalCompilation;
-                return this;
-            }
-
             // run the actual generation
             var state = StateWithPendingEditsApplied(_state);
             var stateBuilder = PooledDictionary<ISourceGenerator, GeneratorState>.GetInstance();
+            var receivers = PooledDictionary<ISourceGenerator, ISyntaxReceiver>.GetInstance();
 
-            //PROTOTYPE: should be possible to parallelize this
             foreach (var generator in state.Generators)
             {
                 // initialize the generator if needed
                 if (!state.GeneratorStates.TryGetValue(generator, out GeneratorState generatorState))
                 {
                     generatorState = InitializeGenerator(generator, cancellationToken);
                 }
+                stateBuilder.Add(generator, generatorState);
 
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    var rx = generatorState.Info.SyntaxReceiverCreator();
+                    receivers.Add(generator, rx);
+                }
+            }
+
+            // Run a syntax walk if any of the generators requested it
+            if (receivers.Count > 0)
+            {
+                GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(receivers.Values.ToImmutableArray());
+                foreach (var syntaxTree in compilation.SyntaxTrees)
+                {
+                    walker.Visit(syntaxTree.GetRoot());
+                }
+            }
+
+            //PROTOTYPE: should be possible to parallelize this
+            foreach (var (generator, generatorState) in stateBuilder.ToImmutableDictionary())
```

Comment:
>ToImmutableDictionary [](start = 69, length = 21)

Consider using `ToArray()` since that should be lighter weight than a dictionary.

Reply:
Could probably just enumerate directly

Reply:
We end up mutating the dictionary with the updated state before storing it, so we need to make a copy either way.

---
In reply to: [395347992](https://github.com/dotnet/roslyn/pull/42507#discussion_r395347992) [](ancestors = 395347992)